server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -826,16 +816,31 @@ private void initializeExecutors()
     }
   }
 
-  private void shutdownExecutors()
+  private void shutdownExecutors() throws InterruptedException
   {
     if (persistExecutor != null) {
       persistExecutor.shutdownNow();
+      Preconditions.checkState(
+          persistExecutor.awaitTermination(365, TimeUnit.DAYS),
```

Comment:
This `awaitTermination` existed before. I changed nothing but just moved into `shutdownExecutors()` because it is called in both `close()` and `closeNow()`.

> also, should we call xxx.shutdownNow() on all executors and then wait so that all of them are stopping in parallel ?

Maybe it's useful if it usually takes so long time. I didn't measure how long it takes.

> also the timeout here is too large.. if something was wrong then some thread would be stuck here potentially indefinitely . I would wait for maybe 5 minutes and then print an error and move on.

It looks that this code was there from the first place. @gianm any thoughts?

Reply:
>> also the timeout here is too large.. if something was wrong then some thread would be stuck here potentially indefinitely . I would wait for maybe 5 minutes and then print an error and move on.
>
> It looks that this code was there from the first place. @gianm any thoughts?

With regard to the huge timeout, when I wrote that I was expecting that some other system would kill tasks that are taking too long to shut down. But I guess it would be fine to print an error after a few minutes and give up.

>> also, should we call xxx.shutdownNow() on all executors and then wait so that all of them are stopping in parallel ?
> 
> Maybe it's useful if it usually takes so long time. I didn't measure how long it takes.

Sure, why not call shutdownNow first and then wait. I don't think it should matter too much.

Reply:
Changed.

Reply:
@jihoonson from the patch it looks like xxx.awaitTermination(..) was added here (also I can't see them in current master https://github.com/druid-io/druid/blob/0105cdbc19828009d21de57a30ba55794a518d30/server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java#L840 )
what am I missing ?

it is called from AppenderatorImpl.closeNow() which should finish asap ... hence the suggestion to add a flag on whether to await or not , then closeNow() wouldn't have await while close() would.

Reply:
It is called outside of `shutdownExecutors()`, but inside of `close()` (https://github.com/druid-io/druid/blob/0105cdbc19828009d21de57a30ba55794a518d30/server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java#L698) and `closeNow()` (https://github.com/druid-io/druid/blob/0105cdbc19828009d21de57a30ba55794a518d30/server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java#L749).

BTW, at first, I moved these awaiting termination codes to inside of `shutdownExecutors()` because `closeNow()` doesn't wait for `pushExecutor` to be terminated in the original code. But, it looks to be intended from the javadoc.

> Do not unlock base persist dir as we are not waiting for push executor to shut down relying on current JVM to shutdown to not cause any locking problem if the task is restored.

I reverted this change and added a comment to avoid confusion.

Reply:
@pjain1 any thoughts?